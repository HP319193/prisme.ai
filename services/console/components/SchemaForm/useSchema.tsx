import { Schema, Tooltip } from '@prisme.ai/design-system';
import { useTranslation } from 'next-i18next';
import { useCallback } from 'react';
import useLocalizedText from '../../utils/useLocalizedText';
import { readAppConfig } from '../AutomationBuilder/Panel/readAppConfig';

function getEmitEvents(
  doList: Prismeai.InstructionList
): Prismeai.Emit['emit'][] {
  return doList.flatMap((instruction) => {
    const [name] = Object.keys(instruction);
    const value = instruction[name as keyof typeof instruction];

    if (Array.isArray(value)) {
      return getEmitEvents(value as Prismeai.InstructionList);
    }
    if (name === 'conditions') {
      return Object.keys(value).flatMap((key) => getEmitEvents(value[key]));
    }
    if (name === 'repeat') {
      return getEmitEvents((value as Prismeai.Repeat['repeat']).do);
    }
    if (name !== 'emit') return [];
    return (instruction as Prismeai.Emit).emit;
  });
}

export const useSchema = (store: Record<string, any> = {}) => {
  const { localize } = useLocalizedText();
  const { t } = useTranslation('workspaces');
  const extractSelectOptions = useCallback(
    (schema: Schema) => {
      const { 'ui:options': uiOptions = {} } = schema;

      switch (uiOptions.from) {
        case 'config':
          const { path = '' } = uiOptions;
          if (!path) return null;
          const values: string[] = readAppConfig(store.config, path);
          return [
            {
              label: '',
              value: '',
            },
            ...values.map((value) => ({
              label: value,
              value,
            })),
          ];
        case 'pageSections':
          return [
            {
              label: '',
              value: '',
            },
            ...(store.pageSections || []).map((sectionId: string) => ({
              label: sectionId,
              value: sectionId,
            })),
          ];
        case 'automations':
          if (!store.automations) return [];
          return [
            {
              label: '',
              value: '',
            },
            ...Object.keys(store.automations).flatMap((key) => {
              const { slug = key, name, description, when } = store.automations[
                key
              ];

              if (
                uiOptions.filter === 'endpoint' &&
                (!when || !when.endpoint)
              ) {
                return [];
              }
              return {
                label: (
                  <div className="flex flex-col">
                    <div>{localize(name) || slug}</div>
                    <div className="text-neutral-500 text-xs">
                      {localize(description)}
                    </div>
                  </div>
                ),
                value: slug,
              };
            }),
          ];
        case 'pages':
          if (!store.pages) return null;
          return [
            {
              label: '',
              value: '',
            },
            ...Array.from<Prismeai.Page>(store.pages).flatMap((page) => {
              const { id, slug, name = slug, description } = page;
              return {
                label: (
                  <div className="flex flex-col">
                    <div>{localize(name)}</div>
                    <div className="text-neutral-500 text-xs">
                      {localize(description)}
                    </div>
                  </div>
                ),
                value: id,
              };
            }),
          ];
      }
      return null;
    },
    [localize, store.automations, store.config, store.pageSections, store.pages]
  );

  const extractAutocompleteOptions = useCallback(
    (schema: Schema) => {
      const { ['ui:options']: uiOptions = {} } = schema;
      const pages: Set<Prismeai.Page> = store.pages || new Set();

      switch (uiOptions.autocomplete) {
        case 'events:listen': {
          const pagesEvents = Array.from(
            pages
          ).flatMap(({ name, blocks = [] }) =>
            blocks.flatMap(({ config: { updateOn } = {} }) =>
              updateOn ? { name: localize(name), event: updateOn } : []
            )
          );

          const apps: Prismeai.AppDetails[] = store.apps || [];
          const automations =
            store?.automations || store.workspace?.automations || {};
          const when: string[] = Object.keys(automations)
            .flatMap((key) => {
              const { events } = (automations[key] || {}).when || {};
              return events || [];
            }, [])
            .filter(Boolean);

          return [
            ...(when.length > 0
              ? [
                  {
                    label: store.workspace.name,
                    options: when.map((event) => ({
                      label: event,
                      value: event,
                    })),
                  },
                ]
              : []),
            ...(pagesEvents.length > 0
              ? [
                  {
                    label: t('pages.link'),
                    options: pagesEvents.map(({ name, event }) => ({
                      label: (
                        <div>
                          {event}
                          <span className="text-[10px] text-neutral-500">
                            {' '}
                            — {name}
                          </span>
                        </div>
                      ),
                      value: event,
                    })),
                  },
                ]
              : []),
            ...apps.flatMap(({ appName, events: { listen = [] } = {} }) =>
              listen.length === 0
                ? []
                : {
                    label: appName,
                    options: (listen || []).map((event) => ({
                      label: event,
                      value: event,
                    })),
                  }
            ),
          ];
        }
        case 'events:emit':
          const workspace = store.workspace as Prismeai.Workspace;
          const apps: Prismeai.AppDetails[] = store.apps || [];
          const automations =
            store?.automations || workspace?.automations || {};

          const events = Object.keys(automations)
            .flatMap((key) => {
              const automation = automations[key];
              if (!automation.do || automation.do.length === 0) return [];
              return getEmitEvents(automation.do);
            })
            // Remove empty and deduplicate
            .filter(
              ({ event }, index, all) =>
                event &&
                !all.slice(0, index).find(({ event: e }) => e === event)
            );

          const pagesEvents = Array.from(
            pages
          ).flatMap(({ name, blocks = [] }) =>
            blocks.flatMap(({ config: { onInit } = {} }) =>
              onInit ? { name: localize(name), event: onInit } : []
            )
          );

          const generateEventsAutomations = (appName: string) => (
            event: string,
            autocomplete: Prismeai.Emit['emit']['autocomplete']
          ) => {
            if (!autocomplete || Object.keys(autocomplete).length === 0)
              return [event];
            // Only read first parameter while the UI is simple
            // To manage many parameters, we would need an autocomplete inside
            // the value autocompleted
            return Object.keys(autocomplete).flatMap((key) => {
              if (!event.match(`{{${key}}}`)) return;
              const template = autocomplete[key].template || '${value}';
              const { from, path } = autocomplete[key];
              if (!from || !path) return [];
              let config = workspace.config || {};
              if (from === 'appConfig') {
                config = workspace.imports?.[appName]?.config || {};
              }
              const values = readAppConfig(config, path).filter(Boolean);
              return (Array.isArray(values)
                ? values
                : [values]
              ).flatMap((value) =>
                template.replace('${value}', event.replace(`{{${key}}}`, value))
              );
            });
          };

          return [
            ...(events.length > 0
              ? [
                  {
                    label: store.workspace.name,
                    options: events
                      .flatMap(({ event, autocomplete }) =>
                        generateEventsAutomations('')(event, autocomplete)
                      )
                      .flatMap((event) => ({
                        label: event,
                        value: event,
                      })),
                  },
                ]
              : []),
            ...(pagesEvents.length > 0
              ? [
                  {
                    label: t('pages.link'),
                    options: pagesEvents.map(({ name, event }) => ({
                      label: (
                        <div>
                          {event}
                          <span className="text-[10px] text-neutral-500">
                            {' '}
                            — {name}
                          </span>
                        </div>
                      ),
                      value: event,
                    })),
                  },
                ]
              : []),
            ...apps
              .filter(({ events: { emit = [] } = {} }) => emit.length > 0)
              .flatMap(
                ({ events: { emit = [] } = {}, appName = '', slug = '' }) => ({
                  label: appName,
                  options: emit.flatMap(({ event, autocomplete }) =>
                    generateEventsAutomations(appName)(
                      event,
                      autocomplete
                    ).flatMap((value) => ({
                      label: value,
                      value: `${slug}.${value}`,
                    }))
                  ),
                })
              ),
          ];
      }

      return [];
    },
    [store]
  );

  return { extractSelectOptions, extractAutocompleteOptions };
};

export default useSchema;
