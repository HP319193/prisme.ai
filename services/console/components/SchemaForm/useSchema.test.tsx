import useSchema from './useSchema';
import renderer from 'react-test-renderer';

it('should fail to build a select from config with empty store', () => {
  let extractSelectOptionsFn: Function = () => null;
  const C = () => {
    const { extractSelectOptions } = useSchema();
    extractSelectOptionsFn = extractSelectOptions;
    return null;
  };
  renderer.create(<C />);
  expect(
    extractSelectOptionsFn({
      type: 'string',
      'ui:widget': 'select',
      'ui:options': {
        from: 'config',
      },
    })
  ).toBeNull();
});

it('should fail to build a select from config without path', () => {
  const config = {
    foo: {
      bar: {
        a: 1,
        b: 2,
        c: 3,
      },
    },
  };
  let extractSelectOptionsFn: Function = () => null;
  const C = () => {
    const { extractSelectOptions } = useSchema({ config });
    extractSelectOptionsFn = extractSelectOptions;
    return null;
  };
  renderer.create(<C />);
  expect(
    extractSelectOptionsFn({
      type: 'string',
      'ui:widget': 'select',
      'ui:options': {
        from: 'config',
      },
    })
  ).toBeNull();
});

it('should build a select from config', () => {
  const config = {
    foo: {
      bar: {
        a: 1,
        b: 2,
        c: 3,
      },
    },
  };
  let extractSelectOptionsFn: Function = () => null;
  const C = () => {
    const { extractSelectOptions } = useSchema({ config });
    extractSelectOptionsFn = extractSelectOptions;
    return null;
  };
  renderer.create(<C />);
  expect(
    extractSelectOptionsFn({
      type: 'string',
      'ui:widget': 'select',
      'ui:options': {
        from: 'config',
        path: 'foo.bar[*]~',
      },
    })
  ).toEqual([
    { label: '', value: '' },
    {
      label: 'a',
      value: 'a',
    },
    {
      label: 'b',
      value: 'b',
    },
    {
      label: 'c',
      value: 'c',
    },
  ]);
});

it('should build a select from pageSections', () => {
  const pageSections = ['foo', 'bar'];
  let extractSelectOptionsFn: Function = () => null;
  const C = () => {
    const { extractSelectOptions } = useSchema({ pageSections });
    extractSelectOptionsFn = extractSelectOptions;
    return null;
  };
  renderer.create(<C />);
  expect(
    extractSelectOptionsFn({
      type: 'string',
      'ui:widget': 'select',
      'ui:options': {
        from: 'pageSections',
      },
    })
  ).toEqual([
    { label: '', value: '' },
    {
      label: 'foo',
      value: 'foo',
    },
    {
      label: 'bar',
      value: 'bar',
    },
  ]);
});

it('should build a select from automations', () => {
  const automations = {
    foo: {
      name: 'Foo',
    },
    bar: {
      name: 'Bar',
      description: 'Bar automation',
      slug: 'do-bar',
    },
  };
  let extractSelectOptionsFn: Function = () => null;
  const C = () => {
    const { extractSelectOptions } = useSchema({ automations });
    extractSelectOptionsFn = extractSelectOptions;
    return null;
  };
  renderer.create(<C />);
  const options = extractSelectOptionsFn({
    type: 'string',
    'ui:widget': 'select',
    'ui:options': {
      from: 'automations',
    },
  });
  expect(options).toEqual([
    { label: '', value: '' },
    {
      label: expect.any(Object),
      value: 'foo',
    },
    {
      label: expect.any(Object),
      value: 'do-bar',
    },
  ]);
  // @ts-ignore
  const labels = renderer.create(options.map(({ label }: any) => label));
  expect(labels).toMatchSnapshot();
});

it('should build a select from endpoint automations ', () => {
  const automations = {
    foo: {
      name: 'Foo',
    },
    bar: {
      name: 'Bar',
      description: 'Bar automation',
      slug: 'do-bar',
      when: {
        endpoint: true,
      },
    },
  };
  let extractSelectOptionsFn: Function = () => null;
  const C = () => {
    const { extractSelectOptions } = useSchema({ automations });
    extractSelectOptionsFn = extractSelectOptions;
    return null;
  };
  renderer.create(<C />);
  const options = extractSelectOptionsFn({
    type: 'string',
    'ui:widget': 'select',
    'ui:options': {
      from: 'automations',
      filter: 'endpoint',
    },
  });

  expect(options).toEqual([
    { label: '', value: '' },
    {
      label: expect.any(Object),
      value: 'do-bar',
    },
  ]);
  // @ts-ignore
  const labels = renderer.create(options.map(({ label }: any) => label));
  expect(labels).toMatchSnapshot();
});

it('should build a select from pages', () => {
  const pages = new Set([
    {
      id: '123',
      name: 'Foo',
    },
    {
      id: '456',
      name: 'Bar',
      description: 'Bar automation',
      slug: 'do-bar',
    },
  ]);
  let extractSelectOptionsFn: Function = () => null;
  const C = () => {
    const { extractSelectOptions } = useSchema({ pages });
    extractSelectOptionsFn = extractSelectOptions;
    return null;
  };
  renderer.create(<C />);
  const options = extractSelectOptionsFn({
    type: 'string',
    'ui:widget': 'select',
    'ui:options': {
      from: 'pages',
    },
  });
  expect(options).toEqual([
    { label: '', value: '' },
    {
      label: expect.any(Object),
      value: '123',
    },
    {
      label: expect.any(Object),
      value: '456',
    },
  ]);
  // @ts-ignore
  const labels = renderer.create(options.map(({ label }: any) => label));
  expect(labels).toMatchSnapshot();
});

it('should autocomplete emit events', () => {
  let extractAutocompleteOptionsFn: Function = () => null;
  const workspace = {
    name: 'workspace',
    imports: {
      'App A': {
        appSlug: 'App A',
        appName: 'App A',
        config: {
          items: {
            foo: {},
            bar: {},
          },
        },
      },
    },
    automations: {
      foo: {
        when: {
          events: ['listen A', 'listen B'],
        },
        do: [
          {
            emit: {
              event: 'emit A',
            },
          },
          {
            wait: {
              timeout: 42,
            },
          },
          {
            emit: {
              event: 'emit B',
            },
          },
        ],
      },
      bar: {
        when: {
          events: ['listen C'],
        },
      },
      empty: {
        do: [
          {
            emit: {
              event: 'emit C',
            },
          },
        ],
      },
    },
  };
  const apps = [
    {
      appName: 'App A',
      slug: 'App A',
      events: {
        emit: [
          {
            event: 'App A emit A',
          },
          {
            event: 'App A emit B {{foo}}',
            source: {
              foo: {
                from: 'appConfig',
                path: 'items[*]~',
              },
            },
          },
        ],
        listen: ['App A listen A'],
      },
    },
    {
      appName: 'App B',
      slug: 'App B',
      events: {
        emit: [
          {
            event: 'App B emit A',
          },
        ],
        listen: ['App B listen A', 'App B listen B'],
      },
    },
  ];
  const C = () => {
    const { extractAutocompleteOptions } = useSchema({
      workspace,
      apps,
    });
    extractAutocompleteOptionsFn = extractAutocompleteOptions;
    return null;
  };
  renderer.create(<C />);

  expect(
    extractAutocompleteOptionsFn({
      'ui:options': {
        autocomplete: 'events:emit',
      },
    })
  ).toEqual([
    {
      label: 'workspace',
      options: [
        {
          label: 'emit A',
          value: 'emit A',
        },
        {
          label: 'emit B',
          value: 'emit B',
        },
        {
          label: 'emit C',
          value: 'emit C',
        },
      ],
    },
    {
      label: 'App A',
      options: [
        {
          label: 'App A emit A',
          value: 'App A.App A emit A',
        },
        {
          label: 'App A emit B foo',
          value: 'App A.App A emit B foo',
        },
        {
          label: 'App A emit B bar',
          value: 'App A.App A emit B bar',
        },
      ],
    },
    {
      label: 'App B',
      options: [
        {
          label: 'App B emit A',
          value: 'App B.App B emit A',
        },
      ],
    },
  ]);
});

it('should autocomplete emit events', () => {
  let extractAutocompleteOptionsFn: Function = () => null;
  const workspace = {
    name: 'workspace',
    imports: {
      'App A': {
        appSlug: 'App A',
        appName: 'App A',
        config: {
          items: {
            foo: {},
            bar: {},
          },
        },
      },
    },
    automations: {
      foo: {
        when: {
          events: ['listen A', 'listen B'],
        },
        do: [
          {
            emit: {
              event: 'emit A',
            },
          },
          {
            wait: {
              timeout: 42,
            },
          },
          {
            emit: {
              event: 'emit B',
            },
          },
        ],
      },
      bar: {
        when: {
          events: ['listen C'],
        },
      },
      empty: {
        do: [
          {
            emit: {
              event: 'emit C',
            },
          },
        ],
      },
    },
  };
  const apps = [
    {
      appName: 'App A',
      slug: 'App A',
      events: {
        emit: [
          {
            event: 'App A emit A',
          },
          {
            event: 'App A emit B {{foo}}',
            source: {
              foo: {
                from: 'appConfig',
                path: 'items[*]~',
              },
            },
          },
        ],
        listen: ['App A listen A'],
      },
    },
    {
      appName: 'App B',
      slug: 'App B',
      events: {
        emit: [
          {
            event: 'App B emit A',
          },
        ],
        listen: ['App B listen A', 'App B listen B'],
      },
    },
  ];
  const C = () => {
    const { extractAutocompleteOptions } = useSchema({
      workspace,
      apps,
    });
    extractAutocompleteOptionsFn = extractAutocompleteOptions;
    return null;
  };
  renderer.create(<C />);

  expect(
    extractAutocompleteOptionsFn({
      'ui:options': {
        autocomplete: 'events:listen',
      },
    })
  ).toEqual([
    {
      label: 'workspace',
      options: [
        {
          label: 'listen A',
          value: 'listen A',
        },
        {
          label: 'listen B',
          value: 'listen B',
        },
        {
          label: 'listen C',
          value: 'listen C',
        },
      ],
    },
    {
      label: 'App A',
      options: [
        {
          label: 'App A listen A',
          value: 'App A listen A',
        },
      ],
    },
    {
      label: 'App B',
      options: [
        {
          label: 'App B listen A',
          value: 'App B listen A',
        },
        {
          label: 'App B listen B',
          value: 'App B listen B',
        },
      ],
    },
  ]);
});

it('should autocomplete events with no value', () => {
  let extractAutocompleteOptionsFn: Function = () => null;
  const workspace = {};
  const apps = [{}];
  const C = () => {
    const { extractAutocompleteOptions } = useSchema({
      workspace,
      apps,
    });
    extractAutocompleteOptionsFn = extractAutocompleteOptions;
    return null;
  };
  renderer.create(<C />);

  expect(
    extractAutocompleteOptionsFn({
      'ui:options': {
        autocomplete: 'events:emit',
      },
    })
  ).toEqual([]);

  expect(
    extractAutocompleteOptionsFn({
      'ui:options': {
        autocomplete: 'events:listen',
      },
    })
  ).toEqual([]);
});

it('should autocomplete nested emit events', () => {
  let extractAutocompleteOptionsFn: Function = () => null;
  const workspace = {
    name: 'workspace',
    automations: {
      foo: {
        do: [
          {
            emit: {
              event: 'level 1',
            },
          },
          {
            conditions: {
              bar: [
                {
                  emit: {
                    event: 'level 2',
                  },
                },
              ],
              default: [
                {
                  conditions: {
                    bar: [
                      {
                        emit: {
                          event: 'level 3',
                        },
                      },
                    ],
                  },
                },
              ],
            },
          },
        ],
      },
      bar: {
        do: [
          {
            all: [
              {
                emit: {
                  event: 'all 1',
                },
              },
              {
                emit: {
                  event: 'all 2',
                },
              },
            ],
          },
        ],
      },
      doo: {
        do: [
          {
            repeat: {
              on: 'boo',
              do: [
                {
                  emit: {
                    event: 'repeat 1',
                  },
                },
              ],
            },
          },
        ],
      },
    },
  };
  const apps = [{}];
  const C = () => {
    const { extractAutocompleteOptions } = useSchema({
      workspace,
      apps,
    });
    extractAutocompleteOptionsFn = extractAutocompleteOptions;
    return null;
  };
  renderer.create(<C />);

  expect(
    extractAutocompleteOptionsFn({
      'ui:options': {
        autocomplete: 'events:emit',
      },
    })
  ).toEqual([
    {
      label: 'workspace',
      options: [
        {
          label: 'level 1',
          value: 'level 1',
        },
        {
          label: 'level 2',
          value: 'level 2',
        },
        {
          label: 'level 3',
          value: 'level 3',
        },
        {
          label: 'all 1',
          value: 'all 1',
        },
        {
          label: 'all 2',
          value: 'all 2',
        },
        {
          label: 'repeat 1',
          value: 'repeat 1',
        },
      ],
    },
  ]);
});
