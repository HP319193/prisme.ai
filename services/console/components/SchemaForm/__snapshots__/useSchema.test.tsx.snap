// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should build a select from automations 1`] = `
Array [
  <div
    className="flex flex-col"
  >
    <div>
      Foo
    </div>
    <div
      className="text-neutral-500 text-xs"
    >
      
    </div>
  </div>,
  <div
    className="flex flex-col"
  >
    <div>
      Bar
    </div>
    <div
      className="text-neutral-500 text-xs"
    >
      Bar automation
    </div>
  </div>,
]
`;

exports[`should build a select from endpoint automations  1`] = `
<div
  className="flex flex-col"
>
  <div>
    Bar
  </div>
  <div
    className="text-neutral-500 text-xs"
  >
    Bar automation
  </div>
</div>
`;

exports[`should build a select from pages 1`] = `
Array [
  <div
    className="flex flex-col"
  >
    <div>
      Foo
    </div>
    <div
      className="text-neutral-500 text-xs"
    >
      
    </div>
  </div>,
  <div
    className="flex flex-col"
  >
    <div>
      Bar
    </div>
    <div
      className="text-neutral-500 text-xs"
    >
      Bar automation
    </div>
  </div>,
]
`;
